conda-store-server/conda_store_server/action/download_packages.py

```diff
@@ -28,14 +43,149 @@ def action_fetch_and_extract_conda_packages(
+                        # PackageCacheData would create the repodata json file
+                        # if it's not present, which would happen in conda-store
+                        # during the lockfile install action. The repodata file
+                        # is not created if it already exists.
+                        #
+                        # The code that does that in conda is similar to the code
+                        # below. However, there is an important difference. The
+                        # code in conda would fail to read the url and return None
+                        # here:
+                        #
+                        #   url = self._urls_data.get_url(package_filename)
+                        #
+                        # And that would result in the channel field of the json
+                        # file being set to "<unknown>". This is a problem because
+                        # the channel is used when querying cache entries, via
+                        # match_individual from MatchSpec, which would always result
+                        # in a mismatch because the proper channel value is
+                        # different.
+                        #
+                        # That would make conda think that the package is not
+                        # available in the cache, so it would try to download it
+                        # outside of this action, where no locking is implemented.
+                        #
+                        # As of now, conda's cache is not atomic, so the same
+                        # dependencies requested by different builds would overwrite
+                        # each other causing random failures during the build
+                        # process.
+                        #
+                        # To avoid this problem, the code below does what the code
+                        # in conda does but also sets the url properly, which would
+                        # make the channel match properly during the query process
+                        # later. So no dependencies would be downloaded outside of
+                        # this action and cache corruption is prevented.
+                        #
+                        # To illustrate, here's a diff of an old conda entry, which
+                        # didn't work, versus the new one created by this action:
+                        #
+                        # --- /tmp/old.txt        2024-02-05 01:08:16.879751010 +0100
+                        # +++ /tmp/new.txt        2024-02-05 01:08:02.919319887 +0100
+                        # @@ -2,7 +2,7 @@
+                        #    "arch": "x86_64",
+                        #    "build": "conda_forge",
+                        #    "build_number": 0,
+                        # -  "channel": "<unknown>",
+                        # +  "channel": "https://conda.anaconda.org/conda-forge/linux-64",
```

Comment:
Did you raise an issue upstream either in conda or conda-lock? I'd like this to be fixed properly at some point so we don't have to carry all this code around.

Reply:
For example, one potential fix would be to tell `conda` to ignore `channel` if set to `<unknown>` (treat it like a wildcard, basically). Another fix would be stop conda-lock from dropping the channel information. I think I've seen this issue in the test suite too when working with conda-libmamba-solver, so it definitely needs a look.

Reply:
Discussed offline. I'll open an issue in the conda-store repo first and then follow up.